File: 20220219/Greedy/오석호_Level2_조이스틱.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

int solution(string name) {
    int answer = 0;
    // 0. 각 알파벳을 만들기 위해 얼마나 이동해야 하는 지 구한다
    for(int i = 0; i < name.size(); ++i){
        int temp = min('Z'+1-name[i],name[i]-'A');
        answer += temp;
    }
    // 1. 왼쪽으로 가야할 지 오른쪽으로 가야할 지 정해야함
    // 왼쪽에 A가 아닌 알파벳이 더 가까운지, 오른쪽이 더 가까운지 확인해서 가까운 쪽으로 이동(greedy)
    // A만 있는 astring에서 name으로 바뀌는 것이랑
    // name에서 astring으로 바뀌는 것이랑 같으므로
    // 반대로 처리해도 같은 결과가 나옴
    // name을 call by value로 불러왔기 때문에
    // 바로 name을 바꿔도 원본에는 변화가 없다
    string astring;
    for(int i = 0; i < name.size(); i++){
        astring += 'A';
    }
    int idx = 0;
    name[0] = 'A';
    while(1){
        // 전부 A면 종료
        if(astring == name)
            return answer;
        
        // 왼쪽으로 이동
        int leftidx = idx;
        int leftcnt = 0;
        
        while(astring[leftidx] == name[leftidx]){
            leftidx--;
            leftcnt++;
            
            // 왼쪽으로 이동시에 예외사항
            if(leftidx == -1)
                leftidx = name.size()-1;
        }
        
        // 오른쪽으로 이동
        int rightidx = idx;
        int rightcnt = 0;
        
        while(astring[rightidx] == name[rightidx]){
            rightidx++;
            rightcnt++;
            
            // 오른쪽으로 이동시에 예외사항
            if(rightidx == name.size()){
                rightidx = 0;
            }      
        }
        
        // 더 적은 횟수를 가는 경우 선택(Greedy)
        // 왼쪽 선택
        if(leftcnt < rightcnt){
            answer += leftcnt;
            idx = leftidx;
        }
        // 오른쪽 선택
        else{
            answer += rightcnt;
            idx = rightidx;
        }
        name[idx] = 'A';
    }

    return answer;
}

File: 20220219/Greedy/오석호_Level2_조이스틱.java
class Solution {
    public int solution(String name) {
//         // 1. A 넣기
//         StringBuilder aString = new StringBuilder("");
//         for (int i = 0; i < name.length(); ++i) aString.append("A");
//         System.out.println(aString);
        
//         // 2. 글자 바꾸기 (굳이 아스키코드 안써도 20글자 이하이므로 완탐해도 될듯)
//         int idx = 0;
//         int cnt = 0;
//         // A = 65, Z = 90
//         // 78까지는 정방향, 90까지는 역방향
//         while(name != aString){
//             if (name.charAt(idx) != aString.charAt(idx)){
//                 if (name.charAt(idx) > 78) cnt += 91-name.charAt(idx);
//                 else cnt += name.charAt(idx) - 65;
//                 aString.setCharAt(idx) = name.charAt(idx);
//             }
//             idx++;
//         }
        // 아.. 진짜 자바는 답이없네
        
        // 1. 글자바꾸기
        int cnt = 0;
        int cursor = name.length() - 1;
        for (int i = 0; i < name.length(); ++i) cnt += name.charAt(i) > 78 ? 91-name.charAt(i) : name.charAt(i) - 65;
        
        // 2. 좌우이동 (완탐)
        for(int i = 0; i <name.length(); i++) {
            int next = i + 1; // 다음 좌표
            while(next != name.length() && name.charAt(next) == 'A') next++;
            
            int As = next - i - 1; // 연속된 A
            int move = name.length() - As - 1; // 전체에서 A를 제외하고 움직이는 횟수
            int min = Math.min(i, name.length() - next); // i까지 무지성 전진 vs next 바로 앞까지 무지성 후진 중 짧은거
            cursor = Math.min(cursor, move + min); // 처음부터 오른쪽만 vs 위에서 짧은거 + 다시 돌아오는거 비교해서 짧은거
        }
        // 각 자리수를 만들기위해서 필요한 조작 횟수 + 좌우를 최소한으로 조작하는 수
        cnt += cursor;
        
        return cnt;
    }
}
